tests/unit/s2n_npn_extension_test.c

```diff
@@ -0,0 +1,198 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include "s2n_test.h"
+
+#include <stdint.h>
+#include <string.h>
+
+#include "tls/s2n_tls.h"
+#include "tls/extensions/s2n_npn.h"
+#include "testlib/s2n_testlib.h"
+#include "tests/s2n_test.h"
+#include "stuffer/s2n_stuffer.h"
+#include "utils/s2n_safety.h"
+
+#define HTTP11 0x68, 0x74, 0x74, 0x70, 0x2f, 0x31, 0x2e, 0x31
+#define SPDY1 0x73, 0x70, 0x64, 0x79, 0x2f, 0x31
+#define SPDY2 0x73, 0x70, 0x64, 0x79, 0x2f, 0x32
+#define SPDY3 0x73, 0x70, 0x64, 0x79, 0x2f, 0x33
+
+int main(int argc, char **argv)
+{
+    BEGIN_TEST();
+
+    const char *protocols[] = { "http/1.1", "spdy/1", "spdy/2" };
+    const uint8_t protocols_count = s2n_array_len(protocols);
+    
+    /* Should-send tests on the client side */
+    {
+        /* No connection */
+        EXPECT_FALSE(s2n_client_npn_extension.should_send(NULL));
+
+        /* No config */
+        DEFER_CLEANUP(struct s2n_connection *client_conn = s2n_connection_new(S2N_CLIENT), s2n_connection_ptr_free);
+        EXPECT_NOT_NULL(client_conn);
+        EXPECT_FALSE(s2n_client_npn_extension.should_send(client_conn));
+
+        /* NPN not supported */
+        DEFER_CLEANUP(struct s2n_config *config = s2n_config_new(), s2n_config_ptr_free);
+        EXPECT_NOT_NULL(config);
+        EXPECT_SUCCESS(s2n_connection_set_config(client_conn, config));
+        EXPECT_FALSE(s2n_client_npn_extension.should_send(client_conn));
+
+        /* NPN supported but no application protocols set */
+        client_conn->config->npn_supported = true;
+        EXPECT_FALSE(s2n_client_npn_extension.should_send(client_conn));
+
+        EXPECT_SUCCESS(s2n_config_set_protocol_preferences(config, protocols, protocols_count));
+        EXPECT_TRUE(s2n_client_npn_extension.should_send(client_conn));
+    }
+
+    /* Should-send tests on the server side */
+    {
+        /* Setup necessary to send NPN */
+        DEFER_CLEANUP(struct s2n_connection *server_conn = s2n_connection_new(S2N_SERVER), s2n_connection_ptr_free);
+        EXPECT_NOT_NULL(server_conn);
+
+        DEFER_CLEANUP(struct s2n_config *config = s2n_config_new(), s2n_config_ptr_free);
+        EXPECT_NOT_NULL(config);
+        EXPECT_SUCCESS(s2n_config_set_protocol_preferences(config, protocols, protocols_count));
+        EXPECT_SUCCESS(s2n_connection_set_config(server_conn, config));
+
+        server_conn->config->npn_supported = true;
+
+        EXPECT_TRUE(s2n_server_npn_extension.should_send(server_conn));
```

Comment:
I know it's pretty trivial, but it is very important: can we verify that should_send is false if npn isn't enabled?